File: stream_of_characters.py

# // Time Complexity :O(n), n is the length of stream
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :no


class TrieNode:
    def __init__(self):
        self.children=[False for i in range(26)]
        self.isEnd=False
class StreamChecker:
    def insert(self,word):
        
        curr=self.root
        for i in range(len(word)-1,-1,-1):
            c=word[i]
            
            if not curr.children[ord(c)-ord('a')]:
                
                curr.children[ord(c)-ord('a')]=TrieNode()
            curr=curr.children[ord(c)-ord('a')]
        curr.isEnd=True
        

    def __init__(self, words: List[str]):
        self.root=TrieNode()
        self.sb=""
        for i in words:
            self.insert(i)
        

    def query(self, letter: str) -> bool:
        curr=self.root
        self.sb+=letter
        
        for i in range(len(self.sb)-1,-1,-1):
            c=self.sb[i]
            
            if not curr.children[ord(c)-ord('a')]:
            
                return False
            if curr.children[ord(c)-ord('a')].isEnd:
                
                return True
            curr=curr.children[ord(c)-ord('a')]
       
            
        


# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)

File: streamOfCharacters.java
/*
Problem: https://leetcode.com/problems/stream-of-characters/
*/
// TC: O(n * k)
// SC: O(n * k)
class TrieNode {
    TrieNode children[];
    boolean isEnd;
    
    public TrieNode() {
        children = new TrieNode[26];
        isEnd = false;
    }
}

class Trie {
    TrieNode root;
    
    public Trie() {
        root = new TrieNode();
    }
    
    public void buildTrie(String words[]) {
        for (String word : words) {
            insert(word);
        }
    }
    
    // Insert word in reverse order
    public void insert(String word) {
        TrieNode cur = root;
        
        for (int i = word.length() - 1; i >= 0; --i) {
            char c = word.charAt(i);
            if (cur.children[c - 'a'] == null) {
                cur.children[c - 'a'] = new TrieNode();
            }
            cur = cur.children[c - 'a'];
        }
        cur.isEnd = true;
    }
}

class StreamChecker {
    Trie t;
    StringBuilder sb;
    
    public StreamChecker(String[] words) {
        sb = new StringBuilder();
        t = new Trie();
        t.buildTrie(words);
    }
    
    public boolean query(char letter) {
        TrieNode cur = t.root;
        sb.append(letter);
        for (int i = sb.length() - 1; i >= 0; --i) {
            char c = sb.charAt(i);
            if (cur.children[c - 'a'] == null) {
                return false;
            }
            
            cur = cur.children[c - 'a'];
            if (cur.isEnd) {
                return true;
            }
        }
        return false;
    }
}

/**
 * Your StreamChecker object will be instantiated and called as such:
 * StreamChecker obj = new StreamChecker(words);
 * boolean param_1 = obj.query(letter);
 */